src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/VirtualDispatchHelpers.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.InteropServices;
+
+namespace System.Runtime.CompilerServices;
+
+[StackTraceHidden]
+[DebuggerStepThrough]
+internal static unsafe partial class VirtualDispatchHelpers
+{
+    private struct VirtualResolutionData : IEquatable<VirtualResolutionData>
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public VirtualResolutionData(MethodTable* methodTable, IntPtr classHandleTargetMethod, IntPtr methodHandle)
+        {
+            HashCode = (int) ((uint)methodTable + (BitOperations.RotateLeft((uint)classHandleTargetMethod, 5)) + (BitOperations.RotateRight((uint)methodHandle, 5)));
+            MethodTable = methodTable;
+            ClassHandleTargetMethod = classHandleTargetMethod;
+            MethodHandle = methodHandle;
+        }
+        public int HashCode;
+        public MethodTable* MethodTable;
+        public IntPtr ClassHandleTargetMethod;
+        public IntPtr MethodHandle;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool Equals(VirtualResolutionData other) =>
+            HashCode == other.HashCode &&
+            MethodTable == other.MethodTable &&
```

Comment:
HashCode should filter out nearly all mismatches, so the rest of the method can assume that the equality check is most likely going to return true. It may help on some architectures to turn the rest of the method to just one conditional branch, something like:
```
        public bool Equals(VirtualResolutionData other) =>
            HashCode == other.HashCode &&
            (((nint)MethodTable - (nint)other.MethodTable) |
            (ClassHandleTargetMethod - other.ClassHandleTargetMethod) |
            (MethodHandle - other.MethodHandle)) == 0;
```